pkg/util/ranger/checker.go

```diff
@@ -168,11 +169,22 @@ func (c *conditionChecker) checkLikeFunc(scalar *expression.ScalarFunction) (isA
 	if err != nil {
 		return false, true
 	}
+	likeFuncReserve := !c.isFullLengthColumn()
+
+	// Different from `=`, trailing spaces are always significant, and can't be ignored in `like`.
+	// In tidb's implementation, for PAD SPACE collations, the trailing spaces are removed in the index key. So we are
+	// unable to distinguish 'xxx' from 'xxx   ' by a single index range scan, and we may read more data than needed by
+	// the `like` function. Therefore, a Selection is needed to filter the data.
+	// Since all collations, except for binary, implemented in tidb are PAD SPACE collations for now, we use a simple
+	// collation != binary check here.
+	if collation != charset.CollationBin {
```

Comment:
What about make PAD SPACE as a new attribute for collation, so that we won't import bugs when a new NO PAD collation is added?

Reply:
Yes. But I think it's better for SQL team to add this into collation-related packages, especially when they start to implement NO PAD collations. I added an `isPadSpaceCollation` function in #48984 since we need the same check there.
This won't cause bugs. If we use current logic when NO PAD collation is added, there will be unnecessary `Selection`, that's not perfect but won't cause bugs.